File: client/src/api/notifications.ts
/**
 * Notifications API client for BaluHost
 */
import { apiClient } from '../lib/api';

// Notification types
export type NotificationType = 'info' | 'warning' | 'critical';
export type NotificationCategory = 'raid' | 'smart' | 'backup' | 'scheduler' | 'system' | 'security' | 'sync' | 'vpn';

export interface Notification {
  id: number;
  created_at: string;
  user_id: number | null;
  notification_type: NotificationType;
  category: NotificationCategory;
  title: string;
  message: string;
  action_url: string | null;
  is_read: boolean;
  is_dismissed: boolean;
  priority: number;
  metadata: Record<string, any> | null;
  time_ago: string | null;
  snoozed_until: string | null;
}

export interface NotificationListResponse {
  notifications: Notification[];
  total: number;
  unread_count: number;
  page: number;
  page_size: number;
}

export interface UnreadCountResponse {
  count: number;
  by_category: Record<string, number> | null;
}

export interface MarkReadResponse {
  success: boolean;
  count: number;
}

export interface CategoryPreference {
  email: boolean;
  push: boolean;
  in_app: boolean;
}

export interface NotificationPreferences {
  id: number;
  user_id: number;
  email_enabled: boolean;
  push_enabled: boolean;
  in_app_enabled: boolean;
  quiet_hours_enabled: boolean;
  quiet_hours_start: string | null;
  quiet_hours_end: string | null;
  min_priority: number;
  category_preferences: Record<string, CategoryPreference> | null;
}

export interface NotificationPreferencesUpdate {
  email_enabled?: boolean;
  push_enabled?: boolean;
  in_app_enabled?: boolean;
  quiet_hours_enabled?: boolean;
  quiet_hours_start?: string | null;
  quiet_hours_end?: string | null;
  min_priority?: number;
  category_preferences?: Record<string, CategoryPreference>;
}

// API Functions

/**
 * Get notifications for the current user
 */
export async function getNotifications(
  options: {
    unread_only?: boolean;
    include_dismissed?: boolean;
    category?: NotificationCategory;
    notification_type?: NotificationType;
    created_after?: string;
    created_before?: string;
    page?: number;
    page_size?: number;
  } = {}
): Promise<NotificationListResponse> {
  const response = await apiClient.get<NotificationListResponse>('/api/notifications', {
    params: {
      unread_only: options.unread_only ?? false,
      include_dismissed: options.include_dismissed ?? false,
      category: options.category,
      notification_type: options.notification_type,
      created_after: options.created_after,
      created_before: options.created_before,
      page: options.page ?? 1,
      page_size: options.page_size ?? 50,
    },
  });
  return response.data;
}

/**
 * Get unread notification count for the current user
 */
export async function getUnreadCount(): Promise<UnreadCountResponse> {
  const response = await apiClient.get<UnreadCountResponse>('/api/notifications/unread-count');
  return response.data;
}

/**
 * Mark a specific notification as read
 */
export async function markAsRead(notificationId: number): Promise<Notification> {
  const response = await apiClient.post<Notification>(`/api/notifications/${notificationId}/read`);
  return response.data;
}

/**
 * Mark all notifications as read
 */
export async function markAllAsRead(category?: NotificationCategory): Promise<MarkReadResponse> {
  const response = await apiClient.post<MarkReadResponse>('/api/notifications/read-all', {
    category,
  });
  return response.data;
}

/**
 * Dismiss a notification
 */
export async function dismissNotification(notificationId: number): Promise<Notification> {
  const response = await apiClient.post<Notification>(`/api/notifications/${notificationId}/dismiss`);
  return response.data;
}

/**
 * Get notification preferences for the current user
 */
export async function getPreferences(): Promise<NotificationPreferences> {
  const response = await apiClient.get<NotificationPreferences>('/api/notifications/preferences');
  return response.data;
}

/**
 * Update notification preferences for the current user
 */
export async function updatePreferences(
  preferences: NotificationPreferencesUpdate
): Promise<NotificationPreferences> {
  const response = await apiClient.put<NotificationPreferences>(
    '/api/notifications/preferences',
    preferences
  );
  return response.data;
}

// Helper functions

/**
 * Get type icon and color for notification type
 */
export function getTypeStyle(type: NotificationType): { icon: string; color: string; bgColor: string } {
  switch (type) {
    case 'critical':
      return {
        icon: '🚨',
        color: 'text-rose-400',
        bgColor: 'bg-rose-500/20 border-rose-500/40',
      };
    case 'warning':
      return {
        icon: '⚠️',
        color: 'text-amber-400',
        bgColor: 'bg-amber-500/20 border-amber-500/40',
      };
    case 'info':
    default:
      return {
        icon: 'ℹ️',
        color: 'text-sky-400',
        bgColor: 'bg-sky-500/20 border-sky-500/40',
      };
  }
}

/**
 * Get category icon
 */
export function getCategoryIcon(category: NotificationCategory): string {
  switch (category) {
    case 'raid':
      return '💾';
    case 'smart':
      return '🔧';
    case 'backup':
      return '📦';
    case 'scheduler':
      return '⏰';
    case 'system':
      return '🖥️';
    case 'security':
      return '🔒';
    case 'sync':
      return '🔄';
    case 'vpn':
      return '🔐';
    default:
      return '🔔';
  }
}

/**
 * Get category display name
 */
export function getCategoryName(category: NotificationCategory): string {
  const names: Record<NotificationCategory, string> = {
    raid: 'RAID',
    smart: 'SMART',
    backup: 'Backup',
    scheduler: 'Scheduler',
    system: 'System',
    security: 'Security',
    sync: 'Sync',
    vpn: 'VPN',
  };
  return names[category] || category;
}

/**
 * Snooze a notification for a given number of hours
 */
export async function snoozeNotification(
  notificationId: number,
  durationHours: number
): Promise<Notification> {
  const response = await apiClient.post<Notification>(
    `/api/notifications/${notificationId}/snooze`,
    null,
    { params: { duration_hours: durationHours } }
  );
  return response.data;
}

/**
 * Get action label for a notification based on its category
 */
export function getActionLabel(category: NotificationCategory): string {
  const labels: Record<NotificationCategory, string> = {
    raid: 'RAID anzeigen',
    smart: 'Disk Health',
    backup: 'Backups anzeigen',
    scheduler: 'Scheduler anzeigen',
    system: 'System anzeigen',
    security: 'Logs anzeigen',
    sync: 'Sync anzeigen',
    vpn: 'VPN anzeigen',
  };
  return labels[category] || 'Anzeigen';
}

/**
 * Fetch a short-lived, scoped token for WebSocket authentication.
 * This avoids passing the full access token as a query parameter.
 */
export async function getWsToken(): Promise<string> {
  const response = await apiClient.post<{ token: string }>('/api/notifications/ws-token');
  return response.data.token;
}

/**
 * Get WebSocket URL for notifications.
 * Token is passed as parameter (single source of truth from AuthContext).
 */
export function getWebSocketUrl(token: string): string {
  const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
  const host = import.meta.env.DEV ? '127.0.0.1:3001' : window.location.host;
  return `${protocol}//${host}/api/notifications/ws?token=${token}`;
}

File: backend/app/api/routes/notifications.py
"""Notification API endpoints."""
from datetime import time, datetime
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Request, Response, status, Query, WebSocket, WebSocketDisconnect
from sqlalchemy.orm import Session

from app.api import deps
from app.core.rate_limiter import user_limiter, get_limit
from app.core.database import get_db
from app.schemas.user import UserPublic
from app.schemas.notification import (
    NotificationCreate,
    NotificationResponse,
    NotificationListResponse,
    UnreadCountResponse,
    MarkReadRequest,
    MarkReadResponse,
    NotificationPreferencesUpdate,
    NotificationPreferencesResponse,
    NotificationCategoryEnum,
)
from app.services.notifications import get_notification_service
from app.services import auth as auth_service
from app.services.permissions import is_privileged

router = APIRouter(prefix="/notifications", tags=["notifications"])


@router.get("", response_model=NotificationListResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def get_notifications(
    request: Request, response: Response,
    unread_only: bool = Query(False, description="Only return unread notifications"),
    include_dismissed: bool = Query(False, description="Include dismissed notifications"),
    category: Optional[NotificationCategoryEnum] = Query(None, description="Filter by category"),
    notification_type: Optional[str] = Query(None, description="Filter by type (info, warning, critical)"),
    created_after: Optional[datetime] = Query(None, description="Only return notifications after this time"),
    created_before: Optional[datetime] = Query(None, description="Only return notifications before this time"),
    page: int = Query(1, ge=1, description="Page number"),
    page_size: int = Query(50, ge=1, le=100, description="Items per page"),
    current_user: UserPublic = Depends(deps.get_current_user),
    db: Session = Depends(get_db),
) -> NotificationListResponse:
    """Get notifications for the current user.

    Returns paginated list of notifications with optional filters.
    """
    service = get_notification_service()
    offset = (page - 1) * page_size
    admin = is_privileged(current_user)

    notifications = service.get_user_notifications(
        db=db,
        user_id=current_user.id,
        unread_only=unread_only,
        include_dismissed=include_dismissed,
        category=category,
        notification_type=notification_type,
        created_after=created_after,
        created_before=created_before,
        limit=page_size,
        offset=offset,
        is_admin=admin,
    )

    unread_count = service.get_unread_count(db, current_user.id, is_admin=admin)

    # Get total count for pagination using SQL COUNT(*)
    total = service.count_user_notifications(
        db=db,
        user_id=current_user.id,
        unread_only=unread_only,
        include_dismissed=include_dismissed,
        category=category,
        notification_type=notification_type,
        created_after=created_after,
        created_before=created_before,
        is_admin=admin,
    )

    return NotificationListResponse(
        notifications=[
            NotificationResponse.from_db(n) for n in notifications
        ],
        total=total,
        unread_count=unread_count,
        page=page,
        page_size=page_size,
    )


@router.get("/unread-count", response_model=UnreadCountResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def get_unread_count(
    request: Request, response: Response,
    current_user: UserPublic = Depends(deps.get_current_user),
    db: Session = Depends(get_db),
) -> UnreadCountResponse:
    """Get count of unread notifications for the current user.

    Also returns breakdown by category.
    """
    service = get_notification_service()

    # Single GROUP BY query instead of 9 separate queries
    counts = service.get_unread_counts(db, current_user.id, is_admin=is_privileged(current_user))
    total = counts.pop("total", 0)

    return UnreadCountResponse(
        count=total,
        by_category=counts if counts else None,
    )


@router.post("/{notification_id}/read", response_model=NotificationResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def mark_notification_as_read(
    request: Request, response: Response,
    notification_id: int,
    current_user: UserPublic = Depends(deps.get_current_user),
    db: Session = Depends(get_db),
) -> NotificationResponse:
    """Mark a specific notification as read."""
    service = get_notification_service()
    notification = service.mark_as_read(db, notification_id, current_user.id, is_admin=is_privileged(current_user))

    if not notification:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Notification not found",
        )

    return NotificationResponse.from_db(notification)


@router.post("/read-all", response_model=MarkReadResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def mark_all_as_read(
    request: Request, response: Response,
    body: Optional[MarkReadRequest] = None,
    current_user: UserPublic = Depends(deps.get_current_user),
    db: Session = Depends(get_db),
) -> MarkReadResponse:
    """Mark all notifications as read.

    Optionally filter by category.
    """
    service = get_notification_service()
    category = body.category if body else None

    count = service.mark_all_as_read(db, current_user.id, category=category, is_admin=is_privileged(current_user))

    return MarkReadResponse(success=True, count=count)


@router.post("/{notification_id}/dismiss", response_model=NotificationResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def dismiss_notification(
    request: Request, response: Response,
    notification_id: int,
    current_user: UserPublic = Depends(deps.get_current_user),
    db: Session = Depends(get_db),
) -> NotificationResponse:
    """Dismiss a notification.

    Dismissed notifications are hidden from the default list but can be retrieved with include_dismissed=true.
    """
    service = get_notification_service()
    notification = service.dismiss(db, notification_id, current_user.id, is_admin=is_privileged(current_user))

    if not notification:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Notification not found",
        )

    return NotificationResponse.from_db(notification)


@router.post("/{notification_id}/snooze", response_model=NotificationResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def snooze_notification(
    request: Request, response: Response,
    notification_id: int,
    duration_hours: int = Query(..., ge=1, le=168, description="Hours to snooze (1-168)"),
    current_user: UserPublic = Depends(deps.get_current_user),
    db: Session = Depends(get_db),
) -> NotificationResponse:
    """Snooze a notification for a given number of hours.

    The notification will be hidden from unread counts and default lists
    until the snooze period expires.
    """
    service = get_notification_service()
    notification = service.snooze(db, notification_id, current_user.id, duration_hours, is_admin=is_privileged(current_user))

    if not notification:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Notification not found",
        )

    return NotificationResponse.from_db(notification)


# Preferences endpoints
@router.get("/preferences", response_model=NotificationPreferencesResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def get_notification_preferences(
    request: Request, response: Response,
    current_user: UserPublic = Depends(deps.get_current_user),
    db: Session = Depends(get_db),
) -> NotificationPreferencesResponse:
    """Get notification preferences for the current user.

    Returns default preferences if none are set.
    """
    service = get_notification_service()
    prefs = service.get_user_preferences(db, current_user.id)

    if not prefs:
        # Create default preferences
        prefs = service.update_user_preferences(db, current_user.id)

    return NotificationPreferencesResponse.from_db(prefs)


@router.put("/preferences", response_model=NotificationPreferencesResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def update_notification_preferences(
    request: Request, response: Response,
    body: NotificationPreferencesUpdate,
    current_user: UserPublic = Depends(deps.get_current_user),
    db: Session = Depends(get_db),
) -> NotificationPreferencesResponse:
    """Update notification preferences for the current user."""
    service = get_notification_service()

    # Parse quiet hours if provided
    quiet_start = None
    quiet_end = None
    if body.quiet_hours_start:
        try:
            parts = body.quiet_hours_start.split(":")
            quiet_start = time(int(parts[0]), int(parts[1]))
        except (ValueError, IndexError):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid quiet_hours_start format. Use HH:MM",
            )
    if body.quiet_hours_end:
        try:
            parts = body.quiet_hours_end.split(":")
            quiet_end = time(int(parts[0]), int(parts[1]))
        except (ValueError, IndexError):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid quiet_hours_end format. Use HH:MM",
            )

    # Convert category preferences if provided
    cat_prefs = None
    if body.category_preferences:
        cat_prefs = {
            cat: pref.model_dump() for cat, pref in body.category_preferences.items()
        }

    prefs = service.update_user_preferences(
        db=db,
        user_id=current_user.id,
        push_enabled=body.push_enabled,
        in_app_enabled=body.in_app_enabled,
        category_preferences=cat_prefs,
        quiet_hours_enabled=body.quiet_hours_enabled,
        quiet_hours_start=quiet_start,
        quiet_hours_end=quiet_end,
        min_priority=body.min_priority,
    )

    return NotificationPreferencesResponse.from_db(prefs)


# Admin endpoints for creating notifications
@router.post("", response_model=NotificationResponse, status_code=status.HTTP_201_CREATED)
@user_limiter.limit(get_limit("admin_operations"))
async def create_notification(
    request: Request, response: Response,
    body: NotificationCreate,
    current_user: UserPublic = Depends(deps.get_current_admin),
    db: Session = Depends(get_db),
) -> NotificationResponse:
    """Create a new notification (admin only).

    Can create notifications for specific users or broadcast to all admins.
    """
    service = get_notification_service()

    notification = await service.create(
        db=db,
        user_id=body.user_id,
        category=body.category,
        notification_type=body.notification_type,
        title=body.title,
        message=body.message,
        action_url=body.action_url,
        metadata=body.metadata,
        priority=body.priority,
    )

    return NotificationResponse.from_db(notification)


# WebSocket token endpoint
@router.post("/ws-token")
@user_limiter.limit(get_limit("user_operations"))
async def get_ws_token(
    request: Request,
    response: Response,
    current_user: UserPublic = Depends(deps.get_current_user),
):
    """Get a short-lived, scoped token for WebSocket authentication.

    The returned token is valid for 60 seconds and only grants access
    to the notification WebSocket endpoint. This avoids passing the
    full access token (which carries broader permissions) as a query
    parameter where it would be logged by proxies and browsers.
    """
    from app.core.security import create_ws_token
    token = create_ws_token(current_user.id, current_user.username)
    return {"token": token}


# WebSocket endpoint for real-time notifications
@router.websocket("/ws")
async def notification_websocket(
    websocket: WebSocket,
    token: Optional[str] = Query(None),
):
    """WebSocket endpoint for real-time notification delivery.

    Requires a scoped WS token (from POST /ws-token) as query parameter.
    Also accepts full access tokens for backwards compatibility (deprecated).
    """
    import logging
    from app.services.websocket_manager import get_websocket_manager
    from app.core.database import SessionLocal
    from app.core.security import decode_token as decode_raw_token
    import jwt as pyjwt

    logger = logging.getLogger(__name__)

    # Validate token before accepting connection
    if not token:
        logger.warning("WebSocket connection attempt without token")
        await websocket.close(code=status.WS_1008_POLICY_VIOLATION)
        return

    # Authenticate user using scoped WS token (preferred) or access token (deprecated)
    user_id = None
    is_admin = False
    db = None

    try:
        db = SessionLocal()

        # First try to decode as a scoped WS token
        try:
            payload = decode_raw_token(token, token_type="ws")
        except pyjwt.InvalidTokenError:
            # Fall back to access token for backwards compatibility
            try:
                payload = decode_raw_token(token, token_type="access")
                logger.warning(
                    "WebSocket: client used access token instead of scoped ws token "
                    "(deprecated, use POST /api/notifications/ws-token)"
                )
            except pyjwt.InvalidTokenError:
                raise auth_service.InvalidTokenError("Invalid WebSocket token")

        user_sub = payload.get("sub")
        from app.services import users as user_service
        user = user_service.get_user(user_sub, db=db)
        if not user:
            logger.warning(f"WebSocket: User not found for token sub={user_sub}")
            await websocket.close(code=status.WS_1008_POLICY_VIOLATION)
            return
        user_id = user.id
        is_admin = user.role == "admin"
        logger.info(f"WebSocket: Authenticated user_id={user_id}, is_admin={is_admin}")
    except auth_service.InvalidTokenError as e:
        logger.warning(f"WebSocket: Invalid token - {e}")
        await websocket.close(code=status.WS_1008_POLICY_VIOLATION)
        return
    except Exception as e:
        logger.error(f"WebSocket: Authentication error - {e}")
        await websocket.close(code=status.WS_1011_INTERNAL_ERROR)
        return
    finally:
        if db:
            db.close()

    # Accept connection after successful authentication
    try:
        await websocket.accept()
        logger.info(f"WebSocket: Connection accepted for user_id={user_id}")
    except Exception as e:
        logger.error(f"WebSocket: Failed to accept connection - {e}")
        return

    # Register with WebSocket manager
    manager = get_websocket_manager()
    await manager.connect(websocket, user_id, is_admin=is_admin)

    try:
        # Send initial unread count with fresh db session
        service = get_notification_service()
        db = SessionLocal()
        try:
            unread_count = service.get_unread_count(db, user_id, is_admin=is_admin)
            await websocket.send_json({
                "type": "unread_count",
                "payload": {"count": unread_count},
            })
        except Exception as e:
            logger.error(f"WebSocket: Failed to send initial unread count - {e}")
        finally:
            db.close()

        # Keep connection alive and handle messages
        while True:
            try:
                data = await websocket.receive_json()

                if data.get("type") == "ping":
                    await websocket.send_json({"type": "pong", "payload": {}})
                elif data.get("type") == "mark_read":
                    notification_id = data.get("payload", {}).get("notification_id")
                    if notification_id:
                        # Use fresh db session for mark_read
                        db = SessionLocal()
                        try:
                            service.mark_as_read(db, notification_id, user_id, is_admin=is_admin)
                            unread_count = service.get_unread_count(db, user_id, is_admin=is_admin)
                            await websocket.send_json({
                                "type": "unread_count",
                                "payload": {"count": unread_count},
                            })
                        except Exception as e:
                            logger.error(f"WebSocket: Failed to mark_read - {e}")
                        finally:
                            db.close()

            except WebSocketDisconnect:
                logger.info(f"WebSocket: Client disconnected user_id={user_id}")
                break
            except Exception as e:
                logger.error(f"WebSocket: Error handling message - {e}")
                break

    finally:
        await manager.disconnect(websocket)
        logger.info(f"WebSocket: Connection closed for user_id={user_id}")
